GUIDE.md

```diff
@@ -1,132 +1,281 @@
+    [ pure PostIndex
+    , Post <$> int
+    , PostEdit <$> int <* lit "edit"
+    , PostBrowse <$> str <*> str
+    ]
 ```
 
-Don't worry too much about the scary type for now.  Essentially, `Match` wraps a function that takes a `Route` (represented as a `List RoutePart`, where `RouteParts` is either the stuff between url slashes or the query parameters) and returns a `V` functor.  `V` stands for 'validation' and you can think of it as an `Either` that can return more than one error upon failure.  So, in essence, something of type `Match Locations` is a function from `Route -> Either Errors (Tuple (Rest-of-RoutePart-List) Locations)`.
-
-The real magic happens in `Match`'s `Apply (i.e., <*>)` implementation.
+This is a lot clearer, but we may have found a bug! Our first route is `pure
+PostIndex`. There are no other conditions to match so this route will always
+succeed, and our subsequent routes won't match. One thing we could do is
+rearrange our routes so that `PostIndex` is last, but that just means
+`PostIndex` will match anything under "posts". What we really want to do is
+match "posts" exactly with no extra path segments. For that we should use the
+`end` combinator.
 
-The `Apply` instance for `Match` provides the plumbing that allows you to compose your routes.  The `<*>`, `*>` and `<*` operators are your friends.  Like any other applicative, `(leftMatch *> rightMatch)` performs the left `Match`, discards a successful result, and returns the value of the right `Match`.  `(leftMatch <* rightMatch)` does the converse.  So, `MPC <$> (homeSlash *> lit "mpc" *> int)` maps the `MPC` constructor function over a `Match` that will parse a slash (and discard it), parse the literal string "mpc" (discarding it too), and return the parsed integer.
+```purescript
+myRoute :: Match MyRoute
+myRoute =
+  lit "posts" *> oneOf
+    [ PostIndex <$ end
+    , Post <$> int <* end
+    , PostEdit <$> int <* lit "edit" <* end
+    , PostBrowse <$> str <*> str <* end
+    ]
+```
 
-Routes are combined using `Match`'s `Alt (<|>)` instance.
+It seems like we might be able to factor out the `end` like we did with `lit
+"posts"`, but that will bring us right back to our bug. It would match any of
+the routes followed by an `end`, so we would still have to rearrange them.
 
 ```purescript
-routing :: Match Locations
-routing =
-  jupiter <|>
-  mars    <|>
-  mpc     <|>
-  moon    <|>
-  home
+myRoute :: Match MyRoute
+myRoute =
+  lit "posts" *> oneOf
+    [ PostEdit <$> int <* lit "edit"
+    , Post <$> int
+    , PostBrowse <$> str <*> str
+    , pure PostIndex
+    ] <* end
 ```
-`jupiter (<|>) mars` means "try Jupiter, and if that fails, try Mars".  Like many routing DSLs, this means that more general routes come after more specific ones.  Thus, `home` (i.e., "/") comes last (unless you only ever want to go home).
 
-We've got routes.  Now what do we do with them?
+We've reduced duplication, but this might be more brittle under refactorings
```

Comment:
At this point I was expecting to read about solution of the ordering issue.
What you think of having a combinator like this `routs = oneOf <<< map ( _ <* end)`

Reply:
I think `<|>` is much more straightforward in comparison with `oneOf`. Introducing `oneOf <<< map (_ <* end)` is too complex for this readme